### PR TITLE
Report const eval error inside the query

### LINKED ISSUE #53561: Constant validation is not cached
Constant validation is not part of the query that evaluates constants, because it would cause cycle errors in case a static directly or transitively contained a pointer back to itself.

One possible solution is to wrap another query around the `const_eval` query, not exposing the `const_eval` query to regular callers (similar to `is_freeze_raw`) and having a wrapper query around it that does the validation and thus also caches the validation.

### PR DESCRIPTION
**Pull request by oli-obk**
Functional changes: We no longer warn about bad constants embedded in unused types. This relied on being able to report just a warning, not a hard error on that case, which we cannot do any more now that error reporting is consistently centralized.

r? @RalfJung 

fixes #53561

### CODE REVIEW

File: src/librustc/dep_graph/dep_node.rs [hunk 1]
```diff
@@ -527,6 +527,7 @@ define_dep_nodes!( <'tcx>
     [] UsedTraitImports(DefId),
     [] HasTypeckTables(DefId),
     [] ConstEval { param_env: ParamEnvAnd<'tcx, GlobalId<'tcx>> },
+    [] ConstEvalRaw { param_env: ParamEnvAnd<'tcx, GlobalId<'tcx>> },
```
Comment (RalfJung) [inline src/librustc/dep_graph/dep_node.rs]: Is there a canonical place to put documentation for a query? The one for this should probably say "DO NOT USE" or so.

File: src/librustc/hir/def.rs [hunk 2]
```diff
@@ -71,6 +71,7 @@ pub enum Def {
     VariantCtor(DefId, CtorKind), // DefId refers to the enum variant
     Method(DefId),
     AssociatedConst(DefId),
+    Closure(hir::BodyId),
```
Comment (RalfJung) [inline src/librustc/hir/def.rs]: I know nothing about HIR. Who could review this part?

File: src/librustc/ty/structural_impls.rs [hunk 3]
```diff
@@ -459,158 +459,6 @@ impl<'a, 'tcx> Lift<'tcx> for ty::error::TypeError<'a> {
...
-            ty::layout::LayoutError::Unknown(ref ty) => {
-                tcx.lift(ty).map(ty::layout::LayoutError::Unknown)
-            }
-            ty::layout::LayoutError::SizeOverflow(ref ty) => {
-                tcx.lift(ty).map(ty::layout::LayoutError::SizeOverflow)
-            }
-        }
-    }
-}
-
```
Comment (RalfJung) [inline src/librustc/ty/structural_impls.rs]: I have no idea why all this code as needed, but I am happy to see it go away :D

File: src/librustc_mir/interpret/memory.rs [hunk 4]
```diff
@@ -142,10 +142,12 @@ fn const_eval_static<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>>(
...
-    tcx.const_eval(ty::ParamEnv::reveal_all().and(gid)).map_err(|err| {
+    // use the raw query here to break validation cycles. Later uses of the static will call the
+    // full query anyway
+    tcx.const_eval_raw(ty::ParamEnv::reveal_all().and(gid)).map_err(|_| {
         // no need to report anything, the const_eval call takes care of that for statics
         assert!(tcx.is_static(def_id).is_some());
-        EvalErrorKind::ReferencedConstant(err).into()
+        EvalErrorKind::ReferencedConstant.into()
     }).map(|val| {
         // FIXME We got our static (will be a ByRef), now we make a *copy*?!?
```
Comment (RalfJung) [inline src/librustc_mir/interpret/memory.rs]: The "copy" part btw is no longer true. `const_to_allocation` asserts that the constant is a `ByRef`, nowadays -- I think. Could you update the comment?

File: src/test/ui/array_const_index-0.stderr [hunk 5]
```diff
@@ -8,5 +8,20 @@ LL | const B: i32 = (&A)[1];
...
    = note: #[deny(const_err)] on by default
 
-error: aborting due to previous error
+error: any use of this value will cause an error
+  --> $DIR/array_const_index-0.rs:12:1
+   |
+LL | const B: i32 = (&A)[1];
+   | ^^^^^^^^^^^^^^^-------^
+   |                |
+   |                index out of bounds: the len is 0 but the index is 1
```
Comment (RalfJung) [inline src/test/ui/array_const_index-0.stderr]: Why do we error twice now for the same thing, where we only errored once before?

**Comment by RalfJung** [comment]
The part inside CTFE/miri looks good to me. I do not think I can adequately review the HIR changes though, so I'd prefer if you could get someone else to take a look at that.

There are a huge amount of changes to the ui test output -- some error messages got deduplicated, but many others (and many more, I think) got duplicated. Why is that? Can you avoid that?

File: src/librustc/ty/structural_impls.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc/ty/structural_impls.rs]: It was needed because queries' args and return value get hashed... but the return type's `Err` does not contain any actual info about the exact error anymore

File: src/test/ui/array_const_index-0.stderr (hunk 5 above)
Comment (oli-obk) [reply src/test/ui/array_const_index-0.stderr]: I presume because the query is once called with `Reveal::UserFacing` and once with `Reveal::All`

File: src/librustc_mir/interpret/memory.rs (hunk 4 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/memory.rs]: I'll address this by nuking the entire query

Comment (oli-obk) [reply src/librustc_mir/interpret/memory.rs]: in a separate PR though

File: src/librustc/dep_graph/dep_node.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc/dep_graph/dep_node.rs]: added. The docs are at the query declaration, not here

File: src/librustc_mir/const_eval.rs [hunk 6]
```diff
@@ -426,6 +427,16 @@ pub fn const_eval_provider<'a, 'tcx>(
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
     key: ty::ParamEnvAnd<'tcx, GlobalId<'tcx>>,
 ) -> ::rustc::mir::interpret::ConstEvalResult<'tcx> {
+    if key.param_env.reveal == Reveal::All {
+        let mut key = key.clone();
+        key.param_env.reveal = Reveal::UserFacing;
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: Please explain in a comment why we even try it with a different than the requested reveal first.

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: Also, why do *both* queries need that logic? Shouldn't one be enough?

File: src/librustc_mir/interpret/place.rs [hunk 7]
```diff
@@ -684,9 +684,16 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
         layout: TyLayout<'tcx>,
         kind: MemoryKind<M::MemoryKinds>,
     ) -> EvalResult<'tcx, MPlaceTy<'tcx>> {
-        assert!(!layout.is_unsized(), "cannot alloc memory for unsized type");
-        let ptr = self.memory.allocate(layout.size, layout.align, kind)?;
-        Ok(MPlaceTy::from_aligned_ptr(ptr, layout))
+        if layout.is_unsized() {
+            assert!(self.tcx.features().unsized_locals, "cannot alloc memory for unsized type");
+            // allocate a fat pointer slot instead
```
Comment (RalfJung) [inline src/librustc_mir/interpret/place.rs]: Wait what? That seems wrong. We can have proper unsized locals, why should we have these indirections? Does that match how the MIR looks?

File: src/librustc_mir/const_eval.rs [hunk 8]
```diff
@@ -427,19 +489,50 @@ pub fn const_eval_provider<'a, 'tcx>(
...
         if tcx.is_static(def_id).is_some() {
-            err.report_as_error(ecx.tcx, "could not evaluate static initializer");
+            let err = err.report_as_error(ecx.tcx, "could not evaluate static initializer");
             if tcx.sess.err_count() == 0 {
-                span_bug!(span, "static eval failure didn't emit an error: {:#?}", err);
+                span_bug!(ecx.tcx.span, "static eval failure didn't emit an error: {:#?}", err);
+            }
+            err
+        } else if def_id.is_local() {
+            // constant defined in this crate, we can figure out a lint level!
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: Could you summarize how the lint level is computed? Seems like constants get a lint (which level does that have?), promotes get error or lint depending on... something, and everything else (statics?) get an error?

File: src/test/ui/array_const_index-0.stderr [hunk 9]
```diff
@@ -8,5 +8,12 @@ LL | const B: i32 = (&A)[1];
    |
    = note: #[deny(const_err)] on by default
 
-error: aborting due to previous error
+error[E0080]: erroneous constant used
+  --> $DIR/array_const_index-0.rs:17:13
+   |
+LL |     let _ = B; //~ ERROR erroneous constant used
+   |             ^ referenced constant has errors
```
Comment (RalfJung) [inline src/test/ui/array_const_index-0.stderr]: So we now error at definition site *and* at use site, but the former is "just" an err-by-default lint? This still complains about the same problem twice, but as long as it's just a lint at definition site I guess this is hard to avoid.

What I do not understand is why behavior changed here, why we now have a hard error in addition to the lint.

File: src/test/ui/consts/const-size_of-cycle.stderr [hunk 10]
```diff
@@ -1,17 +1,27 @@
...
+note: ...which requires normalizing `ParamEnvAnd { param_env: ParamEnv { caller_bounds: [], reveal: All }, value: [u8; _] }`...
+note: ...which requires const-evaluating + checking `Foo::bytes::{{constant}}`...
+  --> $DIR/const-size_of-cycle.rs:16:17
    |
-LL |     unsafe { intrinsics::size_of::<T>() }
-   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^
+LL |     bytes: [u8; std::mem::size_of::<Foo>()]
+   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
+   = note: ...which again requires const-evaluating + checking `Foo::bytes::{{constant}}`, completing the cycle
+note: cycle used when processing `Foo`
```
Comment (RalfJung) [inline src/test/ui/consts/const-size_of-cycle.stderr]: Is it expected that "requires const-evaluating + checking `Foo::bytes::{{constant}}`" occurs *twice* at the end of this cycle?

> ...which requires const-evaluating + checking `Foo::bytes::{{constant}}`...
> ...which again requires const-evaluating + checking `Foo::bytes::{{constant}}`, completing the cycle

File: src/test/ui/recursion/recursive-static-definition.stderr [hunk 11]
```diff
@@ -4,7 +4,17 @@ error[E0391]: cycle detected when const-evaluating `FOO`
...
+  --> $DIR/recursive-static-definition.rs:11:1
+   |
+LL | pub static FOO: u32 = FOO;
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: ...which again requires const-evaluating `FOO`, completing the cycle
+note: cycle used when const-evaluating + checking `FOO`
+  --> $DIR/recursive-static-definition.rs:11:1
+   |
+LL | pub static FOO: u32 = FOO;
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (RalfJung) [inline src/test/ui/recursion/recursive-static-definition.stderr]: Similar here...

> cycle detected when const-evaluating `FOO`
> ...which requires const-evaluating `FOO`...
> ...which again requires const-evaluating `FOO`, completing the cycle

Seems strange to report a 3-element cycle when 2 would suffice?

Comment (RalfJung) [reply src/test/ui/recursion/recursive-static-definition.stderr]: Oh, I guess this is because of the `Reveal` change? Those are different queries that look the same when printed?

File: src/librustc_mir/interpret/place.rs (hunk 7 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/place.rs]: I assumed we'd want to mirror rustc codegen

File: src/test/ui/recursion/recursive-static-definition.stderr (hunk 11 above)
Comment (oli-obk) [reply src/test/ui/recursion/recursive-static-definition.stderr]: yes, that's the issue. We can improve the output to include the environment

File: src/librustc_mir/const_eval.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: if only `const_eval_raw` did it, we might duplicate validation even if the computation of the constant were deduplicated.

if only `const_eval` did it, we might duplicate the computation of the constant in the few explicit calls to `const_eval_raw`.

File: src/test/ui/array_const_index-0.stderr (hunk 9 above)
Comment (oli-obk) [reply src/test/ui/array_const_index-0.stderr]: If you allow the warning, you get an error. All that happened is that the error is reported a little earlier, before the compiler decides to stop compiling further due to "n previous errors"

File: src/librustc_mir/const_eval.rs (hunk 8 above)
Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: done

**Comment by oli-obk** [comment]
If this PR ends up needing rebasing, I won't get to doing that until October 1st. I checked the "allow edits from maintainers" box, so feel free to rebase or apply small fixes to get this landed.

**Comment by alexreg** [comment]
How are we looking with this @RalfJung? I'm happy to finish things off here if any work needs doing while @oli-obk is away.

**Comment by RalfJung** [comment]
I am on vacation, can take a look next week.

**Comment by alexreg** [comment]
Sure, no problem. I'll leave this as a notification for you!

File: src/librustc_mir/interpret/place.rs (hunk 7 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: I don't think so. We already have the ability to use an `Operand::Indirect` with a `MemPlace` to store the metadata -- that's like a properly typed version of a fat pointer.

File: src/librustc_mir/const_eval.rs [hunk 12]
```diff
@@ -367,41 +369,96 @@ pub fn const_variant_index<'a, 'tcx>(
...
+}
+
+pub fn const_eval_raw_provider<'a, 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    key: ty::ParamEnvAnd<'tcx, GlobalId<'tcx>>,
+) -> ::rustc::mir::interpret::ConstEvalResult<'tcx> {
+    // so we do a small trick here. We check whether we can evaluate the constant in the more
+    // restrictive `Reveal::UserFacing`, which most likely already was computed. In a large
+    // percentage of constants that will already have succeeded. Only associated constants of
+    // generic functions will fail due to not enough monomorphization information being available
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: This should say *why* we do this -- to not report the same error twice, because the constant is computed twice (once for each value of `Reveal`).

File: src/test/rustdoc/const-evalutation-ice.rs [hunk 13]
```diff
@@ -17,4 +17,4 @@ pub struct S {
     s: Cell<usize>
 }
 
-pub type _S = [usize; 0 - (mem::size_of::<S>() != 4) as usize];
+pub const N: usize = 0 - (mem::size_of::<S>() != 4) as usize;
```
Comment (RalfJung) [inline src/test/rustdoc/const-evalutation-ice.rs]: Hu, why did this change? Now the comment in line 11 makes no sense any more.

File: src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        // is run on a system whose pointers need more
+        // than 8 bits
+        Bar { a: &42 }.b as u8
+    }
+}
+
+fn main() {
+    // FIXME(oli-obk): this should panic at runtime
+    // this will actually compile, but then
+    // abort at runtime (not panic, hard abort).
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs]: This comment seems rather strange in a test file that does not actually compile.

**Comment by RalfJung** [comment]
@alexreg okay, I think this is mostly good. I added one more nit as a comment.

And then there is the thing about unsized locals. It seems mostly unrelated to this PR, but IIRC tests failed before @oli-obk added that. Ideally we'd find a way to postpone supporting unsized locals in miri...

**Comment by RalfJung** [comment]
Yeah, https://github.com/rust-lang/rust/blob/f1aefb48d2ec7ac38a66c964396a5aec729b7a28/src/test/run-pass-valgrind/unsized-locals/long-live-the-unsized-temporary.rs failed.

Given that we do not actually support unsized locals, I wonder if it is enough to just return `Scalar::ptr_null(cx)` and add a FIXME to do something reasonable later? Could you try that, @alexreg?

File: src/test/rustdoc/const-evalutation-ice.rs (hunk 13 above)
Comment (alexreg) [reply src/test/rustdoc/const-evalutation-ice.rs]: I'm not sure of @oli-obk's exact intention here, but I've resolved it in the way that seems most obvious.

File: src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs (hunk 14 above)
Comment (alexreg) [reply src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs]: Yeah, it's phrased slightly oddly. I think I know what Oli means though, so I've clarified it. :-)

File: src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+union Bar {
+    a: &'static u8,
+    b: usize,
+}
+
+const fn bar() -> u8 {
+    unsafe {
+        // this will error as long as this test
+        // is run on a system whose pointers need more
+        // than 8 bits
```
Comment (alexreg) [inline src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs]: @RalfJung I presume its only runtime behaviour that's dependent on the pointer size here, right? It will always compile?

**Comment by alexreg** [comment]
> Given that we do not actually support unsized locals, I wonder if it is enough to just return `Scalar::ptr_null(cx)` and add a FIXME to do something reasonable later? Could you try that, @alexreg?

Sounds reasonable.

**Comment by RalfJung** [comment]
> Anyway, I pushed my branch to my repo fork.

A link would have been helpful... found your work, but which of the 15 branches there is it?

File: src/test/rustdoc/const-evalutation-ice.rs (hunk 13 above)
Comment (RalfJung) [reply src/test/rustdoc/const-evalutation-ice.rs]: Uh, previously this referred to the `type _S` that oli removed, now you made it refer to the `struct S` instead. That seems like a change in content to me.

Comment (RalfJung) [reply src/test/rustdoc/const-evalutation-ice.rs]: I think what happened is that `_S` was a bogus type so rustdoc ICEd. That got fixed in https://github.com/rust-lang/rust/commit/6b35d813826a8011e3b3f19206c22f6978173134. Now the code actual errors before rustdoc can even start. But we can at least still add a bogus constant and see if rustdoc works on that.

So, I think the comment should refer to `N`, not `S`.

File: src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs (hunk 15 above)
Comment (RalfJung) [reply src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs]: No, this will compile-time-error -- it will fail during "CTFE runtime".

**Comment by RalfJung** [comment]
> Sounds reasonable. What part of the code does this apply to though?

The change in `place.rs`.

We already have `PlaceTy::null`, you could add a `MPlaceTy::null` and use that there, for now, anyway.

File: src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs (hunk 14 above)
Comment (oli-obk) [reply src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs]: Yea this is based on outdated info. feel free to open an issue and assign it to me so I'll investigate and update the comment

File: src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs (hunk 15 above)
Comment (alexreg) [reply src/test/ui/consts/const-eval/promoted_const_fn_fail_deny_const_err.rs]: Hmm, okay. It's a concern then if this test is run on such a machine

File: src/test/rustdoc/const-evalutation-ice.rs (hunk 13 above)
Comment (alexreg) [reply src/test/rustdoc/const-evalutation-ice.rs]: Okay, but `N` isn't a type, so I really wasn't sure what Oli meant...

**Comment by alexreg** [comment]
Yeah, same branch name naturally. Sorry, I thought you could URL-hack it easily. Will look at the remaining issues soon.

**Comment by RalfJung** [comment]
@alexreg I meant here (for `MPlaceTy::null`): https://github.com/rust-lang/rust/pull/53821/files#diff-9c5f25e0d0302fa34791d33795e30781R716

**Comment by RalfJung** [comment]
Hm I think we should make sure this does not affect perf

@bors try

**Comment by RalfJung** [comment]
Never mind, I took care of the unsized local stuff.

@eddyb could you have a look at my commits? I can't r+ this PR alone any more. ;)

File: src/librustc/mir/interpret/mod.rs [hunk 16]
```diff
@@ -156,31 +156,31 @@ impl<'tcx> Pointer {
         Pointer { alloc_id, offset }
     }
 
-    pub fn wrapping_signed_offset<C: HasDataLayout>(self, i: i64, cx: C) -> Self {
+    pub fn wrapping_signed_offset(self, i: i64, cx: impl HasDataLayout) -> Self {
```
Comment (eddyb) [inline src/librustc/mir/interpret/mod.rs]: cc @nikomatsakis I think this looks amazing and we should just change everything throughout!

**Comment by eddyb** [comment]
LGTM, but I only looked at your commits.

**Comment by RalfJung** [comment]
> LGTM, but I only looked at your commits.

Thanks, I took care of the rest. :)

r=me if perf looks good and CI is green.

**Comment by RalfJung** [comment]
@rust-timer build 78ad514b82fc652cb870142fc547ddebe2c5c837

**Comment by RalfJung** [comment]
Lots of green, some red. Looks like solid wins in some places and mild losses in others. Now what do I do with this?^^

**Comment by RalfJung** [comment]
I think the wins are probably because we used to compute the same constant twice, for the two values of `Reveal`?

But the ctfe-stress numbers are ridiculous, it looks like we are not actually doing any work any more there.

**Comment by RalfJung** [comment]
Turns out the lint was not run indeed. Need to re-do perf.

@bors try

File: src/librustc/traits/mod.rs [hunk 17]
```diff
@@ -383,7 +382,7 @@ pub enum SelectionError<'tcx> {
                                 ty::PolyTraitRef<'tcx>,
                                 ty::error::TypeError<'tcx>),
     TraitNotObjectSafe(DefId),
-    ConstEvalFailure(Lrc<ConstEvalErr<'tcx>>),
+    ConstEvalFailure,
```
Comment (eddyb) [inline src/librustc/traits/mod.rs]: This should hold `ErrorHandled`.

**Comment by eddyb** [comment]
cc @nikomatsakis I'm not sure I understand the diagnostic changes.

**Comment by RalfJung** [comment]
The only changes I can see in the tests are removed duplicates, and no longer warning about bad constants embedded in unused type aliases.

@oli-obk also added a new test, `promoted_const_fn_fail_deny_const_err.rs`, which I do not understand -- the comments there make no sense (it claims some code would compile, but the test doesn't compile).

Oh, and validation seems to run on some more things, where previously it didn't run because there were already other errors.

**Comment by eddyb** [comment]
Nominated to discuss diagnostics - I'm leaning towards erroring from constants in type aliases *when* they're monomorphic (which is always right now).

**Comment by pnkfelix** [comment]
discussion has started during [T-compiler meeting](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/subject/weekly.20meeting.202018-09-20/near/134315542) and after the meeting ended the [discussion continued](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/subject/.2353821.20.20Report.20const.20eval.20error.20inside.20the.20query/near/134316202)

I'm leaving the nominated tag on for now because I want to let the main participants post their conclusions here after the discussion is done.

**Comment by RalfJung** [comment]
My summary of the current state is still in https://github.com/rust-lang/rust/pull/53821#issuecomment-423054691. In that diagnostics regression seems acceptable, I think this can land and we can think about producing more errors later.

@eddyb is worried that the approach implemented here is not well suited to just *warn* on bad consts in unused type aliases, should that be what we end up wanting to do. But even then I think we should start from here, and make the queries more flexible, but we should not have `EvalResult` in a query answer -- it is too messy of a type.

**Comment by RalfJung** [comment]
The perf changes are odd. I got both old and new state in perf and callgrind, and the extra cost is hard to pin down. `process_obligations` got most of the extra cost, but that sounds like runs half the compiler... the odd thing is that all the extra cost seems to be in `drop_in_place`?!?

File: src/librustc_mir/hair/cx/expr.rs [hunk 18]
```diff
@@ -572,8 +572,9 @@ fn make_mirror_unadjusted<'a, 'gcx, 'tcx>(cx: &mut Cx<'a, 'gcx, 'tcx>,
             let span = cx.tcx.def_span(def_id);
             let count = match cx.tcx.at(span).const_eval(cx.param_env.and(global_id)) {
                 Ok(cv) => cv.unwrap_usize(cx.tcx),
-                Err(e) => {
-                    e.report_as_error(cx.tcx.at(span), "could not evaluate array length");
+                Err(ErrorHandled::Reported) => 0,
```
Comment (RalfJung) [inline src/librustc_mir/hair/cx/expr.rs]: Should this be a `bug!` or at least `delay_bug`?

**Comment by RalfJung** [comment]
So how do we proceed with this (other than having to wait for @oli-obk because I cannot force-push to his branch)? What's the compiler team's conclusion?

**Comment by RalfJung** [comment]
To summarize (my [comment above](https://github.com/rust-lang/rust/pull/53821#issuecomment-423054691) wasn't entirely clear), this PR as it is right now changes behavior for diagnostics around what happens when a constant fails to evaluate: Is that a hard error, a lint, a warning? Previously, it was up to whatever queried into CTFE to make that decision, with mixed results (and leading to [duplicate errors](https://github.com/rust-lang/rust/issues/52922)). Now it is determined in the query by looking at what the item being evaluated is.

Concretely, it used to be the case that when an erroneous constant was used as array length (or in another way) in a type alias, that caused a *warning* even when the type alias wasn't used -- but caused an *error* if it actually was used. We cannot differentiate like that any more, so to avoid regressing we do not look at unused type aliases any more, they do not warn or error. I have not been able to find any other changes in behavior, but given that previously every call site did whatever it wanted, I may just have missed something. @oli-obk might know more.

So I guess the question is whether that is okay with us. @eddyb also wanted to do some experiments around WF checks for unused type aliases, which is a related question.

**Comment by oli-obk** [comment]
We didn't get to this PR in the compiler meeting. The user visisble changes that this PR causes are summarized in https://github.com/rust-lang/rust/pull/53821#issuecomment-425993073 (and are just "unused type aliases with broken constants in array lengths do not warn anymore").

This PR is a big improvement in our handling of errors around constants implementation wise. You can now simply call `const_eval` and errors/lints will be reported appropriately instead of every call site of the query ad hoc deciding what it wants and duplicating errors or forgetting to report them.

@rfcbot merge

**Comment by nagisa** [comment]
I’m fine with this, but I think we should make a “public” query variant out of `ConstEvalRaw` for the definitely-will-happen scenario where emitting errors from within the query will not be the right thing to do™.

**Comment by alexreg** [comment]
@rust-lang/lang Could we move this forward into FCP stage soon, ideally? :-)

**Comment by Centril** [comment]
@alexreg I think you meant to cc @rust-lang/compiler.

**Comment by alexreg** [comment]
Oops, yes...

**Comment by petrochenkov** [comment]
I know almost nothing about const eval or queries, so I'm ticking my box if RalfJung and eddyb tick theirs.

**Comment by RalfJung** [comment]
I'm not on any team so I don't have a checkbox. But I would tick mine if I did. ;)

**Comment by pnkfelix** [comment]
(removing old and outdated nomination tag)

**Comment by RalfJung** [comment]
> fatal: unable to access 'https://github.com/rust-embedded/cortex-m/'

Seems like Travis' internet doesn't deserve the name, once again :(

@bors retry

**Comment by oli-obk** [comment]
I've seen this failure before locally. Rerunning the tests makes it go away. It's very frustrating. I'll open a tracking issue for it

@bors retry

**Comment by RalfJung** [comment]
What... was that? It started testing twice?!? Cc @rust-lang/infra 

@bors retry

**Comment by nnethercote** [comment]
[Perf results](https://perf.rust-lang.org/compare.html?start=82239b04dc9cea1d54422c2fb223ff5321ccafdd&end=694cf752988728ba5da85c36c2deaeefeb49c05e&stat=instructions:u) are mixed. Some big wins for incremental builds (as high as 27%), and some slowdowns of up to 3% for a few benchmarks, mostly `keccak` and `inflate`.

**Comment by oli-obk** [comment]
We've seen these results before in the PR. attempts to track the regressions down were not very successful. I'll have another look
